zkvm/src/predicate.rs

```diff
@@ -21,25 +21,23 @@ pub enum Predicate {
     /// Verifier's view on the predicate in a compressed form to defer decompression cost.
     Opaque(CompressedRistretto),
 
-    /// Prover's view on the predicate that has valid representation.
-    Witness(PredicateWitness),
-}
-
-/// Prover's representation of the predicate tree with all the secret witness data.
-#[derive(Clone, Debug)]
-pub enum PredicateWitness {
-    /// Representation of an opaque branch of the predicate tree
-    /// (but known to be a valid Ristretto point).
-    OpaqueBranch(RistrettoPoint),
-
-    /// Secret signing key for the predicate-as-a-verification-key.
-    Key(Scalar),
+    /// Signing key for the predicate-as-a-verification-key.
+    /// Prover will provide secret signing key separately when
+    /// constructing aggregated signature.
+    Key(VerificationKey),
 
     /// Representation of a predicate as commitment to a program.
     Program(Vec<Instruction>),
 
     /// Disjunction of two predicates.
-    Or(Box<PredicateWitness>, Box<PredicateWitness>),
+    Or(PredicateDisjunction),
+}
+
+#[derive(Clone, Debug)]
+pub struct PredicateDisjunction {
+    left: Box<Predicate>,
+    right: Box<Predicate>,
+    point: CompressedRistretto,
```

Comment:
clarify that this field is redundant, and comment why it is needed:
```suggestion
    // We precompute the disjunction predicate when composing it via `Predicate::or()`,
    // so that we can keep `to_point`/`encode` methods non-failable across all types.
    precomputed_point: CompressedRistretto,
```